spacy/pipeline/span_finder.py

```diff
@@ -0,0 +1,386 @@
+                    ends.append(token.i)
+
+            for start in starts:
+                for end in ends:
+                    span_length = end + 1 - start
+                    # XXX I really feel like min_length and max_length should be
+                    # None instead of 0 and then just set them to -1 and inf if they
+                    # are given as None.
+                    if span_length > 0:
+                        if (
+                            self.min_length <= 0 or span_length >= self.min_length
+                        ) and (self.max_length <= 0 or span_length <= self.max_length):
+                            doc.spans[self.predicted_key].append(doc[start : end + 1])
+                        elif self.max_length > 0 and span_length > self.max_length:
+                            break
+
+    def update(
+        self,
+        examples: Iterable[Example],
+        *,
+        drop: float = 0.0,
+        sgd: Optional[Optimizer] = None,
+        losses: Optional[Dict[str, float]] = None,
+    ) -> Dict[str, float]:
+        """Learn from a batch of documents and gold-standard information,
+        updating the pipe's model. Delegates to predict and get_loss.
+        examples (Iterable[Example]): A batch of Example objects.
+        drop (float): The dropout rate.
+        sgd (Optional[thinc.api.Optimizer]): The optimizer.
+        losses (Optional[Dict[str, float]]): Optional record of the loss during training.
+            Updated using the component name as the key.
+        RETURNS (Dict[str, float]): The updated losses dictionary.
+        """
+        if losses is None:
+            losses = {}
+        losses.setdefault(self.name, 0.0)
+        predicted = [eg.predicted for eg in examples]
+        set_dropout_rate(self.model, drop)
+        scores, backprop_scores = self.model.begin_update(predicted)
+        loss, d_scores = self.get_loss(examples, scores)
+        backprop_scores(d_scores)
+        if sgd is not None:
+            self.finish_update(sgd)
+        losses[self.name] += loss
+        return losses
+
+    def get_loss(self, examples, scores) -> Tuple[float, Floats2d]:
+        """Find the loss and gradient of loss for the batch of documents and
+        their predicted scores.
+        examples (Iterable[Examples]): The batch of examples.
+        scores: Scores representing the model's predictions.
+        RETURNS (Tuple[float, float]): The loss and the gradient.
+        """
+        reference_truths = self._get_aligned_truth_scores(examples)
+        d_scores = scores - self.model.ops.asarray2f(reference_truths)
+        loss = float((d_scores**2).sum())
+        return loss, d_scores
+
+    def _get_aligned_truth_scores(self, examples) -> List[Tuple[int, int]]:
+        """Align scores of the predictions to the references for calculating the loss"""
+        # TODO: handle misaligned (None) alignments
+        # TODO: handle cases with differing whitespace in texts
```

Comment:
At least the first point is blocking for inclusion in `spacy`. All pipeline components need to be able to handle misaligned tokenization.

For the second point we could potentially error out and say that this just not supported for this component, but it's something that `Example` supports in general and if it's possible to support it it would be strongly preferred. (You can't use plain character offsets at this point, though, so it would be a pretty major refactoring.)

Reply:
I misread the TODOs a bit initially.

The first case is related to returning `None` instead of `0` in cases where the span in the reference document does not align with token boundaries in the predicted doc. In this case, we want to ignore this span for `get_loss` rather than treating it as `0` as is currently the case (I think, anyway, please double-check the details!).

For the second case, we can check that the texts are the same in advance, similar to this in coref:

https://github.com/explosion/spacy-experimental/blob/9dd89d5ed2e4abd160f898c4a4ebae1b297ffb61/spacy_experimental/coref/coref_component.py#L232-L239

Many components in spacy do support training for referenced/predicted docs with differing whitespace, but it's a really niche case and since working from character offsets makes this easier, it's okay to have this constraint here as long as it doesn't accidentally use character offsets for cases where this would be incorrect.

Reply:
I guess I can do it in reverse that makes more sense right? For each `span` in `eg.predicted` i can try to look for corresponding annotation in the `eg.reference` and if I find something I set it to `1` and if I don't I set it to `None`. For all other tokens its `0`.

Reply:
I will push this version, but while we are at it I thought it could be helpful for the model to teach it about the "in" label as well even though it is not used for predictions. This would be a 3rd label used to just tag tokens that are inside a span as "in" for just additional training signal.

Reply:
Then I think this suggestion doesn't really work right?

Reply:
Let me see if I understand everything that's going on here:

- currently the loss function only considers the start and end tokens
- it is a problem that `Doc.char_span` fails if only one is incorrect, since you could still use one of the boundary tokens
  - here maybe you could use `expand` and then filter out cases that aren't the same as the input?
- in `get_loss` should we switch to a loss from thinc that supports missing values?
- you think it might be useful to incorporate the middle tokens so there's more signal for misaligned data but you're not sure this will work?

Reply:
Currently I pushed is what we discussed on Slack. The label assignment decisions are based on the first and last character offsets of the `predicted` and `reference` spans. Using `alignment_mode="expand"` if the first characters and last characters line up then the labeling is what it used to be.

If they don't both line up then the one that lines up gets supervision and the other one is skipped from the loss. This is achieved by masking the token's class specific logit (multiplying the corresponding entry in `d_scores` by 0).